File: csharp/connect/Connect.cs
public enum ConnectWinner
{
    White,
    Black,
    None
}

public class Connect
{
    /* Can move left, right, upper-left, upper-right, lower-left, or lower-right
    O - connect from top to bottom.
    X - connect from left to right.
    */
    // board array (square)
    char[][] board;

    // Possible moves on the hex board
    Tuple<int, int>[] moves = {
        new Tuple<int, int>( -1, -1), new Tuple<int, int>( 1, -1 ),
        new Tuple<int, int>( -2, 0 ), new Tuple<int, int>( 2, 0 ),
        new Tuple<int, int>( -1, 1 ), new Tuple<int, int>( 1, 1 )
    };

    // Make a square board from the given input padding with spaces as needed.
    public Connect(string[] input)
    {
        int height = input.Length;
        int width = (from s in input select s.Length).Max();

        board = (from row in input
                 select (from char c in row.PadRight(width, ' ') select c).ToArray<char>()).ToArray();

    }

    public bool isValidPoint(int x, int y)
    {
        int height = board.Length;
        int width = board[0].Length;
        if (x < 0 || x >= width || y < 0 || y >= height)
            return false;

        return board[y][x] != ' ';
    }

    List<Tuple<int, int>> GetMoves(int x, int y, char player)
    {
        return (from move in moves
         where isValidPoint(x + move.Item1, y + move.Item2) && board[y + move.Item2][x + move.Item1] == player
         select new Tuple<int, int>(x + move.Item1, y + move.Item2)).ToList<Tuple<int, int>>();
    }

    void ClearVisited(bool[][] visitied)
    {
        for (int j = 0; j < visitied.Length; j++)
            for (int i = 0; i < visitied[j].Length; i++)
                visitied[j][i] = false;
    }

    public ConnectWinner Result()
    {
        // Search O - Top to bottom.
        // Get any O on the top row
        Tuple<int, int>[]? candidates = (from pair in board[0].Index()
                                         where pair.Item == 'O'
                                         select new Tuple<int, int>(pair.Index, 0)).ToArray();
        bool[][]? visited = (from row in board
                            select (from ch in row select false).ToArray<bool>()).ToArray();
                            
        if (candidates != null && visited != null)
        {
            ClearVisited(visited);
            foreach (Tuple<int, int> candidate in candidates)
            {
                Stack<Tuple<int, int>> next = new();
                next.Push(candidate);
                while (next.Count > 0)
                {
                    Tuple<int, int> current = next.Pop();
                    visited[current.Item2][current.Item1] = true;

                    // Check if we won
                    if (current.Item2 >= board.Length - 1)
                    {
                        // White made it to bottom side.
                        return ConnectWinner.White;
                    }

                    foreach (Tuple<int, int> item in GetMoves(current.Item1, current.Item2, 'O'))
                    {
                        if (visited[item.Item2][item.Item1] == false)
                        {
                            next.Push(item);
                        }
                    }
                }

            }
        }
        // Search x, left to right.
        candidates = (from row in board.Index()
                      where row.Item[row.Index] == 'X'
                      select new Tuple<int, int>(row.Index, row.Index)).ToArray();

        if (candidates != null && visited != null)
        {
            ClearVisited(visited);
            foreach (Tuple<int, int> candidate in candidates)
            {
                Stack<Tuple<int, int>> next = new();
                next.Push(candidate);
                while (next.Count > 0)
                {
                    Tuple<int, int> current = next.Pop();
                    visited[current.Item2][current.Item1] = true;

                    // Check if we won
                    if (current.Item1 >= board[0].Length - 1 || board[current.Item2][current.Item1 + 2] == ' ')
                    {
                        // X made it to right hand side.
                        return ConnectWinner.Black;
                    }

                    foreach (Tuple<int, int> item in GetMoves(current.Item1, current.Item2, 'X'))
                    {
                        if (visited[item.Item2][item.Item1] == false)
                        {
                            next.Push(item);
                        }
                    }
                }
            }                               
        }
        return ConnectWinner.None;
    }
}

File: go/connect/connect.go
package connect

// Can move left, right, upper-left, upper-right, lower-left, or lower-right
// (this maps counter-intuitively when packed to a rectangle )
//
//	O - connect from top to bottom.
//	X - connect from left to right.
var moves [][2]int = [][2]int{
	{0, -1}, {1, -1},
	{-1, 0}, {1, 0},
	{-1, 1}, {0, 1},
}

// Clear out the visited data resetting all spots to false (not visited)
// @param visited: 2D slice of booleans so we only try each place on the board once.
func ClearVisited(visited [][]bool) {
	for j := 0; j < len(visited); j++ {
		for i := 0; i < len(visited[j]); i++ {
			visited[j][i] = false
		}
	}
}

// Return the possible moves for a position on the board for the given player.
// @param board: slice of string with the board information
// @param width: board width
// @param height: board height
// @param x: x-coordinate of position to search for next move from
// @param y: y-coordinate of position to search for next move from
// @param player: Which player we are looking for moves for.
func GetMoves(board []string, width, height, x, y int, player byte) [][2]int {
	result := make([][2]int, 0)
	for _, move := range moves {
		if IsValidPoint(width, height, x+move[0], y+move[1]) && board[y+move[1]][x+move[0]] == player {
			result = append(result, [2]int{x + move[0], y + move[1]})
		}
	}
	return result
}

// Check if a given point on the board is valid.
// @param width: width of the board
// @param height: height of the board
// @param x: x-coordinate of position to test
// @param y: y-coordinate of position to test
// @returns: True if the position is ok
func IsValidPoint(width, height, x, y int) bool {
	if x < 0 || x >= width || y < 0 || y >= height {
		return false
	}
	return true
}

// Figure out if X has made it from left to right or O from top to bottom on the board
// @param lines: The board represented as a slice of string.
// @returns: "X" = X Wins, "O" = O Wins, "" = No winner.
// @raises: None seems like I should have a test that requires this.
func ResultOf(lines []string) (string, error) {
	height := len(lines)
	width := 0
	for _, row := range lines {
		width = max(width, len(row))
	}

	// Search O - Top to bottom.
	// Get any O on the top row
	var candidates [][2]int = make([][2]int, 0)
	for i, char := range lines[0] {
		if char == 'O' {
			candidates = append(candidates, [2]int{i, 0})
		}
	}
	var visited = make([][]bool, height)
	for j := 0; j < height; j++ {
		visited[j] = make([]bool, width)
		for i := 0; i < width; i++ {
			visited[j][i] = false
		}
	}

	if len(candidates) > 0 {
		ClearVisited(visited)
		for _, candidate := range candidates {
			next := make([][2]int, 0)
			next = append(next, candidate)
			for len(next) > 0 {
				current := next[len(next)-1]
				next = next[:len(next)-1] // Pop
				visited[current[1]][current[0]] = true

				// Check if we won
				if current[1] >= height-1 {
					// White made it to bottom side.
					return "O", nil
				}

				for _, item := range GetMoves(lines, width, height, current[0], current[1], 'O') {
					if visited[item[1]][item[0]] == false {
						next = append(next, item)
					}
				}
			}
		}
	}

	// Search x, left to right.
	candidates = candidates[:0]
	for i, row := range lines {
		if row[0] == 'X' {
			candidates = append(candidates, [2]int{0, i})
		}
	}

	if len(candidates) > 0 {
		ClearVisited(visited)
		for _, candidate := range candidates {
			next := make([][2]int, 0)
			next = append(next, candidate)

			for len(next) > 0 {
				current := next[len(next)-1]
				next = next[:len(next)-1] // Pop
				visited[current[1]][current[0]] = true

				// Check if we won
				if current[0] >= width-1 {
					// X made it to right hand side.
					return "X", nil
				}

				for _, item := range GetMoves(lines, width, height, current[0], current[1], 'X') {
					if visited[item[1]][item[0]] == false {
						next = append(next, item)
					}
				}
			}
		}
	}
	return "", nil
}
